src/backend.rs

```diff
@@ -205,88 +346,639 @@ impl StackDepth {
+        StackValue::Local(loc) => free_value(ctx, Value::Local(loc)),
+        StackValue::Immediate(imm) => free_value(ctx, Value::Immediate(imm)),
+    }
+}
+
+fn pop_i32_into(ctx: &mut Context, dst: ValueLocation) {
+    let val = ctx.block_state.stack.pop().expect("Stack is empty");
+    put_stack_val_into(ctx, val, dst);
+}
+
+fn free_value(ctx: &mut Context, val: Value) {
+    match val {
+        Value::Temp(reg) => ctx.block_state.regs.release_scratch_gpr(reg),
+        Value::Local(_) | Value::Immediate(_) => {}
+    }
+}
+
+/// Puts this value into a register so that it can be efficiently read
+fn into_reg(ctx: &mut Context, val: Value) -> GPR {
+    match val.location(&ctx.block_state.locals) {
+        ValueLocation::Stack(offset) => {
+            let offset = adjusted_offset(ctx, offset);
+            let scratch = ctx.block_state.regs.take_scratch_gpr();
+            dynasm!(ctx.asm
+                ; mov Rq(scratch), [rsp + offset]
+            );
+            scratch
+        }
+        ValueLocation::Immediate(i) => {
+            let scratch = ctx.block_state.regs.take_scratch_gpr();
+            dynasm!(ctx.asm
+                ; mov Rq(scratch), i
+            );
+            scratch
+        }
+        ValueLocation::Reg(reg) => reg,
+    }
+}
+
+/// Puts this value into a temporary register so that operations
+/// on that register don't write to a local.
+fn into_temp_reg(ctx: &mut Context, val: Value) -> GPR {
+    match val {
+        Value::Local(loc) => {
+            let scratch = ctx.block_state.regs.take_scratch_gpr();
+
+            match ctx.block_state.locals.get(loc) {
+                ValueLocation::Stack(offset) => {
+                    let offset = adjusted_offset(ctx, offset);
+                    dynasm!(ctx.asm
+                        ; mov Rq(scratch), [rsp + offset]
+                    );
+                }
+                ValueLocation::Reg(reg) => {
+                    dynasm!(ctx.asm
+                        ; mov Rq(scratch), Rq(reg)
+                    );
+                }
+                ValueLocation::Immediate(_) => {
+                    panic!("We shouldn't be storing immediates in locals for now")
+                }
+            }
+
+            scratch
+        }
+        Value::Immediate(i) => {
+            let scratch = ctx.block_state.regs.take_scratch_gpr();
+
+            dynasm!(ctx.asm
+                ; mov Rq(scratch), i
+            );
+
+            scratch
+        }
+        Value::Temp(reg) => reg,
+    }
+}
+
+macro_rules! commutative_binop {
+    ($name:ident, $instr:ident, $const_fallback:expr) => {
+        pub fn $name(ctx: &mut Context) {
+            let op0 = pop_i32(ctx);
+            let op1 = pop_i32(ctx);
+
+            if let Some(i1) = op1.immediate() {
+                if let Some(i0) = op0.immediate() {
+                    ctx.block_state.stack.push(StackValue::Immediate($const_fallback(i1, i0)));
```

Comment:
Fun, constant folding on the fly :-). Something to consider though: Anyone who cares at all about the performance of their wasm code should have done constant folding ahead of time, because it's trivial to do and a pure code-size/decode-time win (tools like the Emscripten, the LLVM wasm backend, and Binaryen do it, for example). This suggests that the $const_fallback code paths aren't important. Removing them would make this code a little simpler -- admittedly not a lot, but they do add extra code paths that aren't yet covered by tests here (and won't be covered by the spec tests either).

Reply:
You're right, I basically only implemented it because I already wanted immediates for the cases where one argument was an immediate and didn't see any reason _not_ to implement it for the case where both are immediates. I can add a path to all the binop tests that tests with literals if you like, it seems silly to miss this optimisation opportunity since it's so simple to implement.